index.d.ts

```diff
@@ -109,9 +117,11 @@ export interface Unsubscribe {
  * There should only be a single store in a Redux app, as the composition
  * happens on the reducer level.
  *
- * @template S State object type.
+ * @template S The type of state held by this store.
+ * @template A the type of actions which may be dispatched by this store.
+ * @template N The type of non-actions which may be dispatched by this store.
  */
-export interface Store<S> {
+export interface Store<S = {}, A extends Action = Action, N = never> {
```

Comment:
Dispatching non-actions is handled by augmenting the `Dispatch` interface. There's a difference between that and adding another type to the union: for example, when dispatching a thunk, `dispatch` returns the return value of the thunk, not the thunk itself.

Reply:
So are you advocating removal of the `N` parameter?

Reply:
Please don't. Augmenting Dispatch for things like redux-thunk is broken, see https://github.com/gaearon/redux-thunk/issues/82. Still no one gave a solution how typing middlewares can work when redux-thunk has augmented Dispatch.

Reply:
I agree that module augmentation is more of a workaround rather than an ideal solution. 

Still, it gives us something that we can't achieve with just extending the type of `dispatch` argument: it allows us to set up a mapping from input type to output type of `dispatch` calls, which is different from one middleware to another.
Having just `Dispatch<Action | ThunkAction>` is incorrect, because it will work as if `dispatch(thunk)` returned the thunk itself.

Also, from what I see in the issue, the error comes from the definition of `Middleware`, and not from the module augmentation.

Considering the ideal solution: adding `dispatch` signatures happens during store enhancement by `applyMiddleware(...)`. We could try to come up with smarter typings for `StoreEnhancer` and `Middleware` to get signature augmentation without module augmentation. Back then we didn't succeed with it, but now TS is much more powerful.

https://github.com/reactjs/redux/pull/1526
https://github.com/reactjs/redux/issues/1648

Reply:
Maybe the `Dispatch` type should still take 2 type parameters, but one is the input ("action") type, and one is the output type?

Reply:
> Maybe the Dispatch type should still take 2 type parameters, but one is the input ("action") type, and one is the output type?

Sounds good to me.

Reply:
After thinking a bit more about it, I don't think that works. Using a union type for the Action and a union type for the return value does not really help here (e.g. with redux-thunk). What you need instead is two different signatures. I'm thinking the Action shouldn't be the type param (for Store and Middleware) but instead the whole Dispatch should be.